### PR TITLE
implement PlanNode clone interface

### PR DESCRIPTION
**Pull request by czpmango**
PlanNode::clone virtual implementation.

### CODE REVIEW

**Comment by Shylock-Hg** [comment]
Where need the dynamic dispatch?

**Comment by czpmango** [comment]
> Where need the dynamic dispatch?

First, this simplify the invocation:
```
// old invocation
auto* newNode = static_cast<const Limit*>(oldNode)->clone(qctx());
// new invocation
auto* newNode = oldNode->clone();
```
Second, this takes advantage of polymorphism. In many cases, the specific type of PlanNode to be cloned is not cared or known.

**Comment by czpmango** [comment]
> Where need the dynamic dispatch?

Optimizer.

**Comment by Shylock-Hg** [comment]
> > Where need the dynamic dispatch?
> 
> Optimizer.

As I seen, the type of node had known in optimizer.

**Comment by yixinglu** [comment]
> > > Where need the dynamic dispatch?
> > 
> > 
> > Optimizer.
> 
> As I seen, the type of node had known in optimizer.

some patterns don't concern the plan node type of their dependencies, such as: Project -> Any Plan Node

**Comment by Shylock-Hg** [comment]
> > > > Where need the dynamic dispatch?
> > > 
> > > 
> > > Optimizer.
> > 
> > 
> > As I seen, the type of node had known in optimizer.
> 
> some patterns don't concern the plan node type of their dependencies, such as: Project -> Any Plan Node

Ok.Got it.

File: src/optimizer/rule/IndexScanRule.cpp [hunk 1]
```diff
@@ -49,7 +49,9 @@ StatusOr<OptRule::TransformResult> IndexScanRule::transform(OptContext* ctx,
         NG_RETURN_IF_ERROR(createIndexQueryCtx(iqctx, kind, items, qctx, groupNode));
     }
 
-    auto newIN = static_cast<const IndexScan*>(groupNode->node())->clone(qctx);
+    auto* oldIN = groupNode->node();
+    DCHECK_EQ(oldIN->kind(), graph::PlanNode::Kind::kIndexScan);
+    auto* newIN = static_cast<IndexScan*>(oldIN->clone());
```
Comment (Shylock-Hg) [inline src/optimizer/rule/IndexScanRule.cpp:54]: Why remove the `const` modifier.

File: src/planner/Query.h [hunk 2]
```diff
@@ -627,9 +631,13 @@ class Minus final : public SetOp {
         return qctx->objPool()->add(new Minus(qctx, left, right));
     }
 
+    PlanNode* clone() const;
```
Comment (yixinglu) [inline src/planner/Query.h:634]: It's better to add 'override' for this interface.

File: src/planner/Query.h [hunk 3]
```diff
@@ -998,8 +1023,20 @@ class Join : public SingleDependencyNode {
...
+         std::vector<Expression*> probeKeys)
+        : SingleDependencyNode(qctx, kind, input),
+          leftVar_(std::move(leftVar)),
+          rightVar_(std::move(rightVar)),
+          hashKeys_(std::move(hashKeys)),
+          probeKeys_(std::move(probeKeys)) {
+        inputVars_.clear();
+        readVariable(leftVar_.first);
+        readVariable(rightVar_.first);
+    }
```
Comment (yixinglu) [inline src/planner/Query.h:1035]: Why to restore this?

File: src/optimizer/rule/IndexScanRule.cpp (hunk 1 above)
Comment (czpmango) [reply src/optimizer/rule/IndexScanRule.cpp:54]: No changes, oldIN is still const.

File: src/planner/Query.h (hunk 2 above)
Comment (czpmango) [reply src/planner/Query.h:634]: ACK.

File: src/planner/Query.h (hunk 3 above)
Comment (czpmango) [reply src/planner/Query.h:1035]: Just move this ctor to header file.

File: src/optimizer/rule/IndexScanRule.cpp (hunk 1 above)
Comment (Shylock-Hg) [reply src/optimizer/rule/IndexScanRule.cpp:54]: The `const` in static_cast

Comment (czpmango) [reply src/optimizer/rule/IndexScanRule.cpp:54]: newIN is not const, It gonna be modified.

**Review by yixinglu — APPROVED** [review]

**Review by nevermore3 — APPROVED** [review]

**Pull request merged**